### PR TITLE
Refactor notification delivery code

### PR DESCRIPTION
**Pull request by cmrd-senya**
Before this PR notification emiting code was scattered over the code. But there is no reason for that. I refactored the entities' receive/dispatch code so that all the notification emission is in one place. I know, that the old Postzord code will be replaced with some new code based on the federation gem. I think, this refactoring will help to make this work simpler. This PR probably requires some more regression testing than I did by now. I think I'll test it more thoroughly later.

Here are the changes summary:
- `Postzord::Receiver::LocalBatch.notify_users` is picked as a central point for all notifications emission. LocalBatch incapsulates methods that receives a single object with multiple receivers and this logic fits for notification emission. Mention notification method was also moved to the `notify_users`
- There was independent notification code in the Dispatcher. These method were removed and replaced with the `Postzord::Receiver::LocalBatch.notify_users` call. For some reason in the Dispatcher case notification emission relied on a separate worker, while in all other cases it was called inline. Since the Dispatcher works within it's own job anyway, I removed the separate notification worker.
- Method `receive` for relayables returned the saved relayable object while for shareables it returned true/false status. I made shareables return object as well so that the notification method call [here](https://github.com/diaspora/diaspora/compare/develop...cmrd-senya:refactor-local-batch?expand=1#diff-b96a9a9aed59fb685238e5aeafcdfb6cR43) could be used uniformly.
- I removed some tests from the mention and post models test suite, because these tests weren't relevant to the models features.

### CODE REVIEW

**Comment by cmrd-senya** [comment]
Apparently, it also makes the things slower. Will review it against that issue.

**Comment by denschub** [comment]
I'm sorry, Senya, but this PR is absolutely wasted work. @SuperTux88 has refactored large portions of that code already:

* https://github.com/SuperTux88/diaspora/commit/3df1fb83ec9c720c5d829031d35f6b048edb2431
* https://github.com/SuperTux88/diaspora/commit/ebb6d0e6802ffe73a9ad1fc0c6d6ff8ea8b72359

There is no point in merging this and I am not going to merge this since it will cause a lot of work to rebase his work. Sorry, but you are aware of his work and you should really communicate before touching anything regarding the federation layer.

**Comment by cmrd-senya** [comment]
> you should really communicate before touching anything regarding the federation layer.

Communicate? Are you kidding me? Most of my direct questions remains unanswered. I'm not even asking now because I will either be ignored or told something in a manner "we would be more happy without you doing anything, but if you insist you can try to guess our opinion and implement something and receive a reply some time in future, maybe".

Last time I asked @SuperTux88 of his work plan I receive a reply like "I'm going to work on some refactorings". Before, he told be, that he has a solid work plan but ignored me when I asked to share it.
Both of my main tasks include federation changes and that is obvious I have to sync my work with what @SuperTux88 doing. But you both stand like it's only me in only who interested in my productive participation. Of course, I can write you every day begging "please tell me, aren't you by chance working on anything that crosses my work, you great team-member majesty, don't consider me too bothering". But that's not how healthy projects work.

My personal waste of the time is not such a big issue. But your lack of interest in my participation leads to the loss of the paid time of my work. Maybe I'm not as awesome developer as you or @SuperTux88, but I'm capable of doing some useful job and **YOUR** unwillingness to communicate leads to the waste of resources for the diaspora project.

**Comment by denschub** [comment]
Could you please tell me when you wrote a quick "fyi, people, I'm planning to do some refactoring on the notification work in the next weeks" in IRC or in one of our mailing lists? I try to read the backlogs in #diaspora-dev as much as possible, but if it turns out we've missed something, we maybe should think about how we communicate.

Thanks in advance for your corporation,
His Majesty King denschub the Fifth

**Comment by cmrd-senya** [comment]
Ok. Nobody told me that before, but if it is what you expect from me, I'll post on IRC updates on my present work agenda. However, it is not what you and @SuperTux88 personally do, which makes it unfair of you to present the lack of communication as my fault.

Totally depending on your good will,
the bothersome mercenary ill minded vermin, senya.

**Comment by SuperTux88** [comment]
@cmrd-senya sorry, but I thought you were aware of my work. The federation-gem is nearly completed now. So I'm integrating it into diaspora. This means that I rewrite EVERYTHING that has something to do with federation (all `Postzord` stuff, federation-workers, federation-logic in models, and also notifications, because they are called everywhere from federation-code, also all internal local federation logic). Is that what you need to know?

**Comment by cmrd-senya** [comment]
@SuperTux88, thanks. I'm of course aware that the gem integration requires replacing all this ```Postzord``` etc stuff. But it's hard to follow on what stage you are, especially when you work in your fork - I don't see an easy way to monitor changes you make in your personal repository. The same problem with me - I usually write what I intent to work on on my diaspora stream, but @denschub deliberately doesn't want to follow my updates there and he is right in some way - there are better places for the participants coordination. So if you posted somewhere some short updates like "I finished the previous part and pass to the next part" it would be much easier for me to sync my work with your progress. I don't know about the place though, because I'm not on IRC 24/7 now, but that's can be settled.

**Comment by SuperTux88** [comment]
> I don't see an easy way to monitor changes you make in your personal repository.

You can follow them as [rss-feed](https://github.com/SuperTux88/diaspora/commits/federation-gem-salmon.atom) or on diaspora ([all my diaspora work](https://nerdpol.ch/u/commits), [only federation branch](https://nerdpol.ch/tags/diaspora_federation-gem-salmon_push)).

I try to write meaningful commit-messages, so it should be enough when you read the commits. Or should I write what I have done somewhere else?

> I usually write what I intent to work on on my diaspora stream

I don't remember that you wrote something that you intend to refactor notifications, or I would have told you, that I have already done this.

**Comment by Flaburgan** [comment]
Please everybody, it would be lovely if we can stop seeing sarcastic comments. Let's not make things harder than they are. We're facing a single problem: Senya is working on a daily base when reviewers only have a few time to check what's going on. So yes, we need to communicate.

I feel like @SuperTux88 work can already be followed correctly, by checking the issues of his repository which have tasks: https://github.com/SuperTux88/diaspora_federation/issues even if yeah, it's not that easy to have a global picture like "the gem is currently at 80% of the state I target for the 0.6 release of diaspora*". Maybe @SuperTux88 you can comment on https://github.com/diaspora/diaspora/issues/5114 when you push a new significant work? It would be helpful.

On the other side, I also feel like @cmrd-senya is trying to communicate as much as he can, especially because he wants to keep his backers informed. That's maybe not precise enough for the code modification itself though? So maybe @cmrd-senya you can first open an issue before doing a significant work, to say "there is a problem here, I see that improvement possible, is it okay to work on it?". Something as simple would be enough.

Thank you all guys for your work on diaspora*, both is really needed. I'm sure we can find a way to all cooperate efficiently and gently :)

**Comment by cmrd-senya** [comment]
@SuperTux88 
> I try to write meaningful commit-messages, so it should be enough when you read the commits. Or should I write what I have done somewhere else?

I guess this is enough, thanks.

> I don't remember that you wrote something that you intend to refactor notifications, or I would have told you, that I have already done this.

Yes, I didn't write about that because it's not something big, I started this spontaneously while I was prepairing the implementation of notifications for #6818 (mentions in comments). It took not much time anyway, so it's not a big loss. If it was anything that required some days of work, I would inform on that separately. I guess, as your work is nearly finished, I can rebase my work on your federetaion-gem-salmon branch to work with the newest up-to-date code and don't face the old code issues.

@Flaburgan 

> So maybe @cmrd-senya you can first open an issue before doing a significant work, to say "there is a problem here, I see that improvement possible, is it okay to work on it?".

That was exactly what I did a few times before. Check #6780 #6806 #6811. I'm doing this for every significant problem I discover. As you may see, it doesn't receive too much response. Yes, the labels were added to the issues, so it was read by @denschub. But, for example, #6780 says "please, give me feedback on the idea" and I suppose it means more than a label assignment. So after that I started working on #6783 without any approval, because I can't wait forever and this is what I wanted to have for the implemenation of the testsuite for #6750. Of course, after not receiving any response I start to doubt whether I should post such proposals or PRs is the only form of communication I can really receive some feedback with. What is the point of asking questions, if most of the time I'm being ignored?

**Comment by SuperTux88** [comment]
> I guess, as your work is nearly finished, I can rebase my work on your federetaion-gem-salmon branch to work with the newest up-to-date code and don't face the old code issues.

I don't know if the current state is ok to work with, everything is pretty broken now :wink: 
* receive from remote pod works with the new code (including notifications on receive)
* local receive doesn't work (partially broken)
* send should still work with the old code, didn't test this.
* relayables don't relay

If you are aware of this brokenness you can rebase, but I won't recommend it at the current state, if you want to test stuff.

But you are welcome to review the new parts and give feedback on it.

**Comment by Flaburgan** [comment]
>What is the point of asking questions, if most of the time I'm being ignored?

I raised this point [in the new loomio discussion](https://www.loomio.org/d/9H5P0zKB/project-management-at-diaspora-or-what-the-heck-am-i-doing-here-). Your opinion is welcome, let's discuss about this problem there and not on github ;)

**Comment by cmrd-senya** [comment]
> If you are aware of this brokenness you can rebase, but I won't recommend it at the current state, if you want to test stuff.

The main question is whether my work will be merged before or after your gem integration. This is very important and that's strange to raise it so late and in this place, but it must be discussed. If my work is going to be merged before your, than there is no point for me to rebase. Personally, I think the AccountMigration message #6750 must be merged as early as possible, so more software versions can support the migration (even if there is no UI for that, it will support the migration message from the newer pods properly). At the same time, I don't know your terms of your plan and I also don't know how my work will be accepted. So maybe the real moment when my changes can be merged lays after your gem integration. That's important to figure out for me to move further and to avoid some more rejected work. #1851 is facing the same issue, but it's less urgent, it can wait for the gem integration, so I could probably rebase it.

@Flaburgan, thank you for sharing the link, I would miss the discussion otherwise, because I wasn't subscribed on the "governance" group and got no notification of that.

**Comment by cmrd-senya** [comment]
@SuperTux88, maybe it's possible to backport your notification changes upon the current develop branch of diaspora? I guess the notification changes doesn't depend too much on how the federation is implemented? For me it'll help to implement notifications for #6818. I also think that accepting changes by smaller sets helps to notice if something gets broken and to review the change set, so it could be useful for you, because your change set from your federation-gem-salmon branch will become smaller relatively to the develop branch. If this idea is approved as worthy, I can do the job of the backporting.

**Comment by SuperTux88** [comment]
I don't think it is worth to backport some changes to the old federation. The new federation works completely different than the old logic.
